### ISSUE TITLE
NanoAODEDMEventContent

### ISSUE THREAD

**Issue by veelken** [cms-nanoAOD/cmssw#115]
Hi,

I think we need to modify PhysicsTools/NanoAOD/python/NanoAODEDMEventContent_cff.py such that all the branches that we want to store in the nanoAOD are explicitly mentioned, separate for data and MC.

The issue is that when running the nanoAOD production on data, I get a run-time exception that is due to accessing some information which exists only for MC. If I understand correctly, the exception is due to an incompatibility of the PAT tools (e.g. PhysicsTools/PatAlgos/python/tools/jetTools.py), which require cmsRun to be run in unscheduled mode, the presence of modules that are attached to the process object and must not be run for data (because they require information that exists only for MC, e.g. generator-level objects), and the presence of "keep *" statements in NanoAODEDMEventContent_cff.py, which cause these modules to be run.

Compatibility with the PAT tools is very important for recreating PAT objects on the fly (e.g. running jet algorithms not stored on miniAOD, updating tau ID discriminators, etc). The best solution seems to me to modify NanoAODEDMEventContent_cff.py such that all branches that we want to store in the nanoAOD are explicitly mentioned, separate for data and MC, making cmsRun safe to run in unscheduled mode.

**Comment by gpetruc** [comment]
There's no `keep *`.
The only things that are kept with a wildcard are TriggerResults (which are produced by the framework) or NanoAOD specific objects.
So, this is not supposed to interfere with PAT Jet tools etc.

Can you be more specific on the problem you are observing?

**Comment by veelken** [comment]
Hi Giovanni, 

here is the specific error message:

Begin processing the 1st record. Run 303838, Event 2129701824, LumiSection 1820 at 19-Feb-2018 16:17:07.454 EET %MSG-e TauGenJetProducer: TauGenJetProducer:tauGenJets 19-Feb-2018 16:17:20 EET Run: 303838 Event: 2129701824 cannot get collection: InputTag: label = prunedGenParticles, instance = %MSG %MSG-w UndefinedPartonMode: HadronAndPartonSelector:patJetPartons 19-Feb-2018 16:17:20 EET Run: 303838 Event: 2129701824 Could not automatically determine the hadronizer type and set the correct parton selection mode. Parton-based jet flavour will not be defined. %MSG ----- Begin Fatal Exception 19-Feb-2018 16:17:20 EET----------------------- An exception of category 'ProductNotFound' occurred while [0] Processing Event run: 303838 lumi: 1820 event: 2129701824 stream: 0 [1] Running path 'end' [2] Prefetching for module NanoAODOutputModule/'out' [3] Prefetching for module GlobalVariablesTableProducer/'ttbarCategoryTable' [4] Prefetching for module GenTtbarCategorizer/'categorizeGenTtbar' [5] Calling method for module GenHFHadronMatcher/'matchGenCHadron' Exception Message: Principal::getByToken: Found zero products matching all criteria Looking for type: reco::JetFlavourInfoMatchingCollection Looking for module label: slimmedGenJetsFlavourInfos Looking for productInstanceName: Additional Info: [a] If you wish to continue processing events after a ProductNotFound exception, add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration. ----- End Fatal Exception ------------------------------------------------- 19-Feb-2018 16:17:21 EET Closed file file:/hdfs/local/ram/SingleElectron_Run2017E_17Nov2017-v1_MINIAOD_LeptonFakeRateSync_files/CE96B3D0-90FB-E711-9DB4-002590A80E1E.root

**Comment by gpetruc** [comment]
Unscheduled execution only runs modules that are added to a task in the
process, and nothing is added by default to a task in the NanoAOD code.
So, it seems that in your cfg to be you're adding MC modules either to a
path in the schedule or to a task in the process.
Either way, it's your cfg that needs fixes, not the NanoAOD event content


On Mon, Feb 19, 2018 at 4:29 PM, Christian Veelken <notifications@github.com

**Comment by veelken** [comment]
Hi Giovanni,

the problem was indeed that too many modules (some of which needed generator-level information) were attached to the process object and added to Tasks. The problem was caused by a tool developed by the JetMET group ( https://github.com/HEP-KBFI/tth-nanoAOD/blob/master/python/jetToolbox_cff.py#L955-L968 ). I have contacted the authors of the tool about this.

I think it would still be good to define the event content of the nanoAOD by explicit keep statements, similar to what is done for the other official CMS data tiers (defined in Configuration/EventContent/python/EventContent_cff.py). This will add robustness against this type of error where too many modules are attached to the process objects and added to Tasks.

Cheers,

Christian
